ql/src/java/org/apache/hadoop/hive/ql/parse/StorageFormat.java

```diff
@@ -81,11 +84,37 @@ public boolean fillStorageFormat(ASTNode child) throws SemanticException {
       }
       break;
     case HiveParser.TOK_STORAGEHANDLER:
-      storageHandler = processStorageHandler(child.getChild(0).getText());
-      if (child.getChildCount() == 2) {
-        BaseSemanticAnalyzer.readProps(
-          (ASTNode) (child.getChild(1).getChild(0)),
-          serdeProps);
+      for (int i = 0; i < child.getChildCount(); i++) {
+        ASTNode grandChild = (ASTNode) child.getChild(i);
+        switch (grandChild.getToken().getType()) {
+          case HiveParser.TOK_FILEFORMAT_GENERIC:
+            HiveStorageHandler handler;
+            try {
+              handler = HiveUtils.getStorageHandler(conf, storageHandler);
+            } catch (HiveException e) {
+              throw new SemanticException("Failed to load storage handler:  " + e.getMessage());
+            }
+
+            String fileFormatPropertyKey = handler.getFileFormatPropertyKey();
+            if (fileFormatPropertyKey != null) {
+              String fileFormat = grandChild.getChild(0).getText();
+              if (serdeProps.containsKey(fileFormatPropertyKey)) {
+                throw new SemanticException("Provide only one of the following: STORED BY " + fileFormat +
+                    " or WITH SERDEPROPERTIES('" + fileFormatPropertyKey + "'='" + fileFormat + "')");
+              }
+
+              serdeProps.put(fileFormatPropertyKey, fileFormat);
+            } else {
+              throw new SemanticException("STORED BY is not supported for storage handler " +
```

Comment:
Shouldn't this be `STORED AS is not supported..`?

Reply:
Yepp,  typo